lib/dec/sse42.c

```diff
@@ -0,0 +1,72 @@
+// If we have SSE4.2 support, pick off 16 bytes at a time for as long as we can,
+// but make sure that we quit before seeing any == markers at the end of the
+// string. Also, because we write four zeroes at the end of the output, ensure
+// that there are at least 6 valid bytes of input data remaining to close the
+// gap. 16 + 2 + 6 = 24 bytes:
+while (srclen >= 24)
+{
+	// Load string:
+	__m128i str = _mm_loadu_si128((__m128i *)c);
+
+	// The input consists of six character sets in the Base64 alphabet,
+	// which we need to map back to the 6-bit values they represent.
+	// There are three ranges, two singles, and then there's the rest.
+	//
+	//  #  From       To        Add    Index  Characters
+	//  1  [43]       [62]      +19        0  +
+	//  2  [47]       [63]      +16        1  /
+	//  3  [48..57]   [52..61]   +4  [2..11]  0..9
+	//  4  [65..90]   [0..25]   -65       15  A..Z
+	//  5  [97..122]  [26..51]  -71       14  a..z
+	// (6) Everything else => invalid input
+
+	// LUT:
+	const __m128i lut = _mm_setr_epi8(19,16,4,4,4,4,4,4,4,4,4,4,0,0,-71,-65);
+
+	// Ranges to be checked:
+	const __m128i range = _mm_setr_epi8('+','+','+','+','+','+','+','+','/','/','0','9','A','Z','a','z');
+
+	// Check for invalid input:
+	if (_mm_cmpistrc(range, str, _SIDD_UBYTE_OPS | _SIDD_CMP_RANGES | _SIDD_NEGATIVE_POLARITY)) {
```

Comment:
I feel that this deserves a bit longer comment to explain what's going on and what the result will be.